File: packages/quran-engine-py/quran_engine/tajweed.py
"""
Tajweed coloring by consuming the pre-computed annotation corpus (data/tajweed/).

This is strategy (A) from docs/PORTING.md: load annotations + map each rule to a color. It needs no
detection logic, so it's small and exactly consistent with the reference engine. To color arbitrary text
(other qiraat, user input) you'd port the detector from docs/02-tajweed.md instead.
"""
from __future__ import annotations
from .models import TajweedSpan
from .text import utf16_slice


class Tajweed:
    def __init__(self, annotations_by_key: dict[tuple[int, int], list[dict]], colors_by_rule: dict[str, str]):
        self._ann = annotations_by_key
        self._colors = colors_by_rule

    def spans(self, surah_id: int, ayah_id: int, ayah_text: str) -> list[TajweedSpan]:
        out = []
        for a in self._ann.get((surah_id, ayah_id), []):
            start, end, rule = a["start"], a["end"], a["rule"]
            out.append(TajweedSpan(
                start=start, end=end, rule=rule,
                text=utf16_slice(ayah_text, start, end),
                color=self._colors.get(rule),
            ))
        return out

File: packages/quran-engine-kotlin/src/main/kotlin/com/quranengine/Tajweed.kt
package com.quranengine

/**
 * Tajweed coloring via strategy (A) from `docs/PORTING.md`: consume the pre-computed annotation corpus
 * (`data/tajweed/NNN.json` / `tajweed-annotations.json`) and map each `rule` to its `colorHex` from
 * `tajweed-rules.json`. No detection logic, so it is small and exactly consistent with the reference.
 *
 * Annotation `start`/`end` are UTF-16 code-unit offsets. Kotlin/JVM `String` is UTF-16 internally, so
 * `text.substring(start, end)` slices on the same units the reference engine uses — no conversion needed.
 */
class Tajweed(
    private val annotationsByKey: Map<Pair<Int, Int>, List<TajweedAnnotation>>,
    private val colorsByRule: Map<String, String>,
) {
    /** Resolve the colored spans for an ayah given its (UTF-16) text. */
    fun spans(surahId: Int, ayahId: Int, ayahText: String): List<TajweedSpan> {
        val anns = annotationsByKey[surahId to ayahId] ?: return emptyList()
        return anns.map { a ->
            // start/end are UTF-16 offsets; JVM String is UTF-16, so substring is a direct slice.
            val safeEnd = minOf(a.end, ayahText.length)
            val safeStart = a.start.coerceIn(0, safeEnd)
            TajweedSpan(
                start = a.start,
                end = a.end,
                rule = a.rule,
                text = ayahText.substring(safeStart, safeEnd),
                colorHex = colorsByRule[a.rule],
            )
        }
    }
}
